vignettes/markdown.Rmd

```diff
@@ -1,17 +1,188 @@
-<!--
-%\VignetteEngine{knitr::knitr}
-%\VignetteIndexEntry{Write R documentation in Markdown}
--->
+---
+title: "Write R Documentation in Markdown"
+author: "Gábor Csárdi"
+date: "`r Sys.Date()`"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Write R Documentation in Markdown}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
 
-```{r, echo = FALSE, message = FALSE}
-knitr::opts_chunk$set(
-  comment = "#>",
-  error = FALSE,
-  tidy = FALSE
-)
+```{r, include = FALSE}
+knitr::opts_chunk$set(comment = "#>", collapse = TRUE)
 ```
 
-# Write R documentation in Markdown
+# Introduction
 
-This is a vignette to be written once we know what exactly the Markdown
-support will cover.
+Starting from version 6.0.0, roxygen supports markdown markup within most roxygen tags. Roxygen uses the `commonmark` package, which is based on the CommonMark Reference Implementation to parse these tags. See http://commonmark.org/help/ for more about the parser and the markdown language it supports.
+
+# Turning on markdown support
+
+Currently you have to turn on markdown formatting manually for each roxygen chunk. You can do this by using the `@md` tag anywhere in the chunk. In the future roxygen will provide a global switch that turns on markdown formatting for the whole package. (And also a `@noMd` to turn it off, optionally.)
+
+Here is an example roxygen chunk that uses markdown.
+
+```r
+#' Use roxygen to document a package.
+#'
+#' This function is a wrapper for the [](roxygen2::roxygenize) function from 
```

Comment:
Preprocessing could be possible, but it could also be fragile. E.g. you would not want to pre-process within code, and there could be other exceptions as well. Even just determining whether you are inside code or not would basically need a markdown parser....

I think, however, that the following could work to get the `[pkg::fun]` and also `[fun]` notation:
1. parse the text, extract the `[pkg::fun]` and `[fun]` forms. All of them, even if they are in code, or wherever.
2. add them to the end of the text, after two newlines, so that cmark can use them as anchors. E.g. `[pkg::fun]: dummy`. 
3. the destination must be some random marker, instead of the `dummy` here, and then when see the marker in the parsed xml, we'll know that this is a link to R documentation.
4. run the parser
5. find the `<link>` tags with the random marker as the `destination`, and replace them with `\code{\link{...}}`.

The unused reference definitions at the end are just ignored, so if we picked up something from a code block, nothing happens:

```r
❯ cat(markdown_xml("foo [pkg::fun] bar `[code]` bar bar\n\n[pkg::fun]: dummy\n[code]: dummy2"))
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE document SYSTEM "CommonMark.dtd">
<document xmlns="http://commonmark.org/xml/1.0">
  <paragraph>
    <text>foo </text>
    <link destination="dummy" title="">
      <text>pkg::fun</text>
    </link>
    <text> bar </text>
    <code>[code]</code>
    <text> bar bar</text>
  </paragraph>
</document>
```